File: typescript/design_pattern/singleton/conventional_singleton.ts
class Database {
	private static instance: Database;

	private constructor() {}

	public static getInstance(): Database {
		if (!Database.instance) {
			Database.instance = new Database()
		}

		return Database.instance;
	}
}

function main() {
	const db1 = Database.getInstance();
	const db2 = Database.getInstance();

	console.log(db1 === db2)
}

main()

File: java/src/main/java/com/example/design_pattern/singleton/ConventionalSingleton.java
package com.example.design_pattern.singleton;

public class ConventionalSingleton {

  private static class Database {
    private static Database db;

    private Database() {}

    public static Database getInstance() {
      if (db == null) {
        db = new Database();
      }
      return db;
    }
  }

  public static void main(String... args) {
    Database db1 = Database.getInstance();
    Database db2 = Database.getInstance();

    assert db1 == db2;
  }
}
